### DISCUSSION TITLE
Performance bottlenecks with `crew_controller_local()` and ~32k dynamic branches

### DISCUSSION THREAD

**Discussion by philiporlando** [wlandau/crew.cluster#47] [Q&A]
I'm experiencing some performance bottlenecks when trying to use `crew_controller_local()` on a target with ~32k dynamic branches. My workflow involves downloading ~32k source files from an AWS S3 bucket and tracking changes using `format = "file"`. 

The main issue is that RAM usage on my server gradually accumulates over time, which eventually slows down the processing to the point where running everything single-threaded, without a controller, is actually faster. Each individual branch runs fairly efficiently, completing in around 0.2 seconds, with minimal CPU or memory usage per branch, so I suspect the overall RAM consumption is related to overhead from managing such a large number of dynamic branches with {crew}.

To mitigate the issue, I've enabled `garbage_collection = TRUE` in `crew_controller_local()`, `tar_option_set()`, and `tar_config_set()`, but the RAM buildup problem persists. 

I'm aware of the [overhead concerns related to creating numerous targets](https://books.ropensci.org/targets/performance.html#overhead), but in this case, tracking changes for each individual S3 objects and local file is critical, making it difficult to abstract the work into fewer, higher-level targets as recommended here:

> Dynamic branching makes it easy to create many targets. Unfortunately, if the number of targets exceeds several thousand, overhead may build up and the package may slow down. Temporary workarounds can avoid overhead in specific cases: for example, the shortcut argument of tar_make(), and choosing a pattern like slice() or head() instead of a full map(). But to minimize overhead at scale, it is better to accomplish the same amount of work with a fewer number of targets. In other words, do more work inside each dynamic branch.

Has anyone else experienced similar memory issues with large-scale dynamic branching or found effective strategies to manage memory consumption in this context? Any suggestions or insights would be greatly appreciated!

----

This issue is similar to https://github.com/ropensci/targets/discussions/1329, but since I’m not using Slurm, I thought it might be worth creating a separate post. Feel free to close this if you believe it’s more appropriate to address in the existing thread.

**Comment by philiporlando** [comment]
This duplicates https://github.com/ropensci/targets/discussions/1347.

**Reply by wlandau** [reply]
Continuing the discussion in https://github.com/ropensci/targets/discussions/1347#discussioncomment-11009365
